### PR TITLE
CFY-4679 Added waiting for elasticsearch shards

### PR DESCRIPTION
**Pull request by xwombat**

### CODE REVIEW

File: components/elasticsearch/scripts/create.py [hunk 1]
```diff
@@ -238,6 +238,24 @@ def _install_elasticsearch():
...
+    url = '{ip}:{port}/*/_search_shards'.format(ip=ip, port=port)
+    while True:
+        all_shards_started = True
+        proc = utils.run(['curl', '-XGET', url])
+        for shard in json.loads(proc.aggr_stdout)['shards']:
+            all_shards_started = all_shards_started and \
+                              (shard[0]['state'] == 'STARTED')
+            if all_shards_started:
+                return
+            time.sleep(1)
```
Comment (tehasdf) [inline components/elasticsearch/scripts/create.py]: indentation?

File: components/elasticsearch/scripts/create.py [hunk 2]
```diff
@@ -238,6 +238,24 @@ def _install_elasticsearch():
...
 
 
+def _wait_for_shards(port, ip):
+    ctx.logger.info('Waiting for shards to be active...')
+    shards_check_timeout = 60
+    start = time.time()
+    url = '{ip}:{port}/*/_search_shards'.format(ip=ip, port=port)
+    while True:
+        all_shards_started = True
+        proc = utils.run(['curl', '-XGET', url])
```
Comment (nir0s) [inline components/elasticsearch/scripts/create.py]: please don't use curl. use urllib/2 instead.

File: components/elasticsearch/scripts/create.py [hunk 3]
```diff
@@ -238,6 +238,24 @@ def _install_elasticsearch():
...
+        for shard in json.loads(proc.aggr_stdout)['shards']:
+            all_shards_started = all_shards_started and \
+                              (shard[0]['state'] == 'STARTED')
+        if all_shards_started:
+            return
+        time.sleep(1)
+        if time.time() - start > shards_check_timeout:
+            utils.error_exit('Elasticsearch shards check timeout')
```
Comment (nir0s) [inline components/elasticsearch/scripts/create.py]: should be "timed out". Also, can we possible get any information about the reason from the rest call?

File: components/elasticsearch/scripts/create.py [hunk 4]
```diff
@@ -238,6 +238,24 @@ def _install_elasticsearch():
     utils.systemd.enable(ES_SERVICE_NAME, append_prefix=False)
 
 
+def _wait_for_shards(port, ip):
+    ctx.logger.info('Waiting for shards to be active...')
```
Comment (nir0s) [inline components/elasticsearch/scripts/create.py]: please add a docstring explaining the reason behind this. it's not obvious why we need this.

File: components/elasticsearch/scripts/create.py [hunk 5]
```diff
@@ -238,6 +238,46 @@ def _install_elasticsearch():
...
+            ctx.abort_operation('Failed to retrieve information about '
+                                'Elasticsearch shards: {0}'.format(e.reason))
+
+        if all_shards_started:
+            return
+        time.sleep(1)
+        if time.time() - start > shards_check_timeout:
+            not_active = [s[0] for s in shards if s[0]['state'] != 'STARTED']
+            ctx.abort_operation('Elasticsearch shards check timed out. '
+                                'Not active shards: {0}'.format(not_active))
```
Comment (nir0s) [inline components/elasticsearch/scripts/create.py]: not_active -> inactive

File: components/elasticsearch/scripts/create.py [hunk 6]
```diff
@@ -238,6 +238,46 @@ def _install_elasticsearch():
...
+    url = 'http://{ip}:{port}/*/_search_shards'.format(ip=ip, port=port)
+    while True:
+        all_shards_started = True
+        try:
+            out = urllib2.urlopen(url)
+            shards = json.load(out)['shards']
+            for shard in shards:
+                all_shards_started = all_shards_started and \
+                                  (shard[0]['state'] == 'STARTED')
+        except urllib2.URLError as e:
```
Comment (nir0s) [inline components/elasticsearch/scripts/create.py]: I'm wondering.. how often does this happen? might it actually fail even if it's only a matter of time until elasticsearch is query-able?

Comment (tehasdf) [inline components/elasticsearch/scripts/create.py]: At this point, elasticsearch is necessarily up and running - function that run before `_wait_for_shards` (ie. `_configure_elasticsearch`) already send requests to it.

**Pull request merged**